syntax = "proto3";

package aeraki.meta_protocol_proxy.filters.ratelimit.v1alpha;

import "envoy/config/ratelimit/v3/rls.proto";

import "google/protobuf/duration.proto";
import "udpa/annotations/status.proto";
import "validate/validate.proto";

option java_package = "io.aeraki.meta_protocol_proxy.filters.ratelimit.v1alpha";
option java_outer_classname = "RateLimitProto";
option java_multiple_files = true;
option (udpa.annotations.file_status).package_version_status = ACTIVE;

message RateLimit {
  // items
  repeated RateLimitItem items = 1;
  // service
  envoy.config.ratelimit.v3.RateLimitServiceConfig rate_limit_service = 2 [(validate.rules).message = {required: true}];
}

message RateLimitItem {
  // domain
  string domain = 1 [(validate.rules).string = {min_len: 1}];

  // timeout in ms
  uint32 timeout = 2;

  // failure deny
  bool failure_mode_deny = 3;

  repeated RateLimitMatch match = 4;
}

message RateLimitMatch {
  // key
  string key = 1 [(validate.rules).string = {min_len: 1}];
  // value match, empty to match all
  string value = 2;
}
